File: prob03/python/prob03.py
from __future__ import absolute_import

# begin: kludge to include ../eulertoolspy/functionalpy
import os
import sys
eulerpath = os.path.abspath('..')
if eulerpath not in sys.path:
    sys.path = [eulerpath]+sys.path
# end: kludge

k = 600851475143

# we know p^2 <= x

def biggestFactor(n):
    candidate = 2
    factor = 1
    # we will keep reducing n by going [n,n/f1, (n/f1)/f2, ..., 1] as
    # we keep updating our latest biggest factor f1,f2,f3 start with n
    while n > 1:
        # if 'factor' is really a factor ...
        if n % candidate == 0:
            # ... remember it as a real factor
            factor=candidate
            # reduce n by that factor, as much as possible
            n = n / factor
            while n % factor ==0:
                n = n / factor
        # try the next possible factor
        candidate += 1
    return factor

def primeFactors(x):
    """Calculates all primes of a number.

    This algorithm is the sieve of eratothsenes. But requires a
    collection the size of the number, so no good for large numbers"""
    import math
    candidates = list(range(2,x))
    primes = []
    while(len(candidates) > 0):
        candidate = candidates.pop(0)
        print "trying candidates = ", candidate
        # if candidate is a prime ..
        if x % candidate == 0:
            print " found a prime ", candidate
            # .. record that and purge its multiples
            primes.append(candidate)
            candidates = [i for i in candidates if i % candidate != 0]

    return primes


def primeFactorsMap(x):
    """Returns a map of all prime factors of x to their powers in x"""
    import math
    factors = {}
    # would rather use list of primes here, instead of integers
    candidates = list(range(2,x))
    while x > 1:
        candidate = candidates.pop(0)
        print "trying candidates = ", candidate
        # if candidate is a factor ...
        if x % candidate == 0:
            print "found a factor ", candidate
            # purge its multiples
#            candidates = [i for i in candidates if i % candidate != 0]

            if candidate not in factors:
                factors[candidate]=0

            while x % candidate ==0:
                # ... record that and purge its multiples
                factors[candidate] += 1
                x = x / candidate
    return factors
            


          
          
          
def isPrime(x):
    """ Checks primality via 'trial division'.

    Makes O(n) attempted divisions"""
    import math
    for i in xrange(2,1+int(math.sqrt(x))):
#        print "Trying factor i=", i
        if x % i == 0:
            return False
    return True

def prime_factors(num, factor=2):
    """Return all prime factors of num in an ordered list"""
    from eulertoolspy.functionalpy import chain, first
    from math import sqrt
    
    if num <= 1:
        return []
    # We know sqrt(num)+1 is the upper bound.
    # a stream of candidate, starting with 'factor'
    # E.g.,   [factor,factor+1,...,sqrt(num)+1,num]
    candidates = chain(xrange(factor, int(sqrt(num))+1), [num])
    # the next actual prime factor is the next candidate factor that is
    # actually a factor
    next = first(x for x in candidates if (num%x == 0))
    # now that we've found next is an actual factor of num, we can
    # divide num/next and look for more factors greater than or equal
    # to next.
    return [next] + prime_factors(num/next, next)
 
answer  = max(prime_factors(k))

File: prob03/cpp/prob03.cpp
#include <iostream>
#include <map>
#include <list>

// boost provides 64 bit integers to non-gcc compilers
#include <boost/cstdint.hpp>
using boost::uint64_t;

using namespace std;

/* Populates factors with the powers of n's prime factors */
void factorInteger(uint64_t n, map<uint64_t,unsigned int> &factors) 
{
  // start trying candidate factors
  for(uint64_t candidate=2; n > 1 && candidate^2 < (1+n^2); ++candidate) {

    // initialise factors to 0
    if(n % candidate == 0) {
      factors[candidate]=0;
    }

    // if candidate is a factor ..
    while(n % candidate == 0) {
      // then factor out all its multiples
      n = n / candidate;
      factors[candidate] += 1;
    }
  }

  return;
}


/* Returns the largest prime factor of n. 

   This function takes [2,3,...,inf] as a list of candidate factors,
   and one by one tries to divide them out of n as many times as is
   possible. Since it tries repeatedly on a factor, it will
   automatically divide out powers of that factor. Thus, it will only
   divide out primes. Since the candidates are listed in order, the
   last factor to divide out successfuly will be the highest prime
   factor of n.
*/
long biggestFactor(uint64_t n)
{
  // the largest factor so far
  long factor = 1;
  // a candidate factor
  long candidate = 2;

  // as long as n is not completely factored ...
  while(n > 1) {
    // ... try our next candidate factor
    if(n % candidate == 0) {
      // if it is a factor, remember it as the highest so far,
      factor = candidate;

      //  and divide it out of n
      n = n / factor;
      while(n % factor == 0) {
	n = n / factor;
      }
    }
    // then prepare to try the next highest factor
    candidate += 1;
  }
  return factor;
}

/* Returns a list of all prime factors of n. 

   This function takes [2,3,...,inf] as a list of candidate factors,
   and one by one tries to divide them out of n as many times as is
   possible. Since it tries repeatedly on a factor, it will
   automatically divide out powers of that factor. Thus, it will only
   divide out primes. Since the candidates are listed in order, the
   last factor to divide out successfuly will be the highest prime
   factor of n.
*/
template<typename T>
std::list<unsigned long> factorIntegerList(T n)
{
  // the prime factors so far
  std::list<unsigned long> factors;
  // a candidate factor
  unsigned long candidate = 2;

  // as long as n is not completely factored ...
  while(n > 1) {
    // ... try our next candidate factor
    if(n % candidate == 0) {
      // if it is a factor, remember it as the highest so far,
      factors.push_back(candidate);

      //  and divide it out of n
      n = n / candidate;
      while(n % candidate == 0) {
	n = n / candidate;
      }
    }
    // then prepare to try the next highest factor
    candidate += 1;
  }
  return factors;
}


// prints a map
template<typename K,typename V>
void printmap(std::map<K,V> mmap) {
  clog << "   map = ";
  clog << "{ ";

  for(typename map<K,V>::iterator thisitem_iter = mmap.begin();
      thisitem_iter != mmap.end(); 
      ++thisitem_iter) {
    clog << thisitem_iter->first << ":" << thisitem_iter->second << ", ";
  }
  clog << "}" << endl;
}

// prints a list
template<typename T>
void printlist(std::list<T> lst) {
  clog << "   list = ";
  clog << "[ ";

  for(typename list<T>::iterator thisitem_iter = lst.begin();
      thisitem_iter != lst.end();
      ++thisitem_iter) {
    clog << *thisitem_iter << ", "; 
  }

  clog << "]" << endl;
}

int main()
{
  uint64_t k = 600851475143LL;
  long factor = biggestFactor(k);
  std::cout << "largest factor of " << k << " is " << factor << std::endl;

  std::map<uint64_t,unsigned int> m;
  factorInteger(k,m);
  printmap(m);

  std::list<unsigned long> lst;
  lst.push_back(1);
  lst.push_back(2);
  printlist(lst);

  printlist(factorIntegerList(k));


  std::cout << "sizeof char* = " << sizeof(char*) << std::endl;
  std::cout << "sizeof long = " << sizeof(long) << std::endl;
}

